File: app/src/main/java/com/dpis/module/templates/QuickTemplateSaveHandler.java
package com.dpis.module.templates;

import com.dpis.module.viewport.ViewportTargetType;

import com.dpis.module.R;

import java.util.LinkedHashSet;
import java.util.Set;

public final class QuickTemplateSaveHandler {
    public Result save(QuickTemplateStore store, Request request) {
        if (store == null) {
            return Result.failure(R.string.status_save_requires_init, null);
        }
        if (request == null) {
            return Result.failure(R.string.status_save_invalid, null);
        }
        String name = request.name.trim();
        if (name.isEmpty()) {
            return Result.failure(R.string.quick_template_name_required, request.templateId);
        }
        if (store.hasDuplicateName(name, request.templateId)) {
            return Result.failure(R.string.quick_template_name_duplicate, request.templateId);
        }
        if (!TemplateCustomSemantics.isViewportInputValid(
                request.viewportInput, request.viewportTargetType)
                || !TemplateCustomSemantics.isFontScaleInputValid(request.fontScaleInput)) {
            return Result.failure(R.string.status_save_invalid, request.templateId);
        }

        String templateId = request.templateId;
        if (templateId == null || templateId.trim().isEmpty()) {
            templateId = store.newTemplateId();
        }
        QuickTemplateStore.QuickTemplate currentTemplate = store.read(templateId);
        Set<String> selectedPackages = currentTemplate != null
                ? currentTemplate.selectedPackages
                : new LinkedHashSet<>();
        TemplateConfigValue configValue = buildConfigValue(request);
        QuickTemplateStore.QuickTemplate template = new QuickTemplateStore.QuickTemplate(
                templateId,
                name,
                System.currentTimeMillis(),
                selectedPackages,
                configValue);
        return store.save(template)
                ? Result.success(R.string.quick_template_save_success, templateId)
                : Result.failure(R.string.quick_template_save_failed, templateId);
    }

    private static TemplateConfigValue buildConfigValue(Request request) {
        return TemplateCustomSemantics.fromEditorDraft(
                request.viewportInput,
                request.viewportTargetType,
                request.viewportApplyMode,
                request.viewportScaleInput,
                request.viewportAbsoluteInput,
                request.fontScaleInput,
                request.fontApplyMode,
                request.selectedTypefaceId,
                request.fontHookDomainsRaw);
    }

    public static final class Request {
        public final String templateId;
        public final String name;
        public final String viewportInput;
        public final String viewportTargetType;
        public final String viewportApplyMode;
        public final String viewportScaleInput;
        public final String viewportAbsoluteInput;
        public final String fontScaleInput;
        public final String fontApplyMode;
        public final String selectedTypefaceId;
        public final String fontHookDomainsRaw;

        public Request(String templateId,
                String name,
                String viewportInput,
                String viewportTargetType,
                String viewportApplyMode,
                String fontScaleInput,
                String fontApplyMode,
                String selectedTypefaceId,
                String fontHookDomainsRaw) {
            this(templateId, name, viewportInput, viewportTargetType, viewportApplyMode,
                    TemplateCustomSemantics.draftInputForTargetType(
                            viewportInput, viewportTargetType,
                            TemplateConfigValue.VIEWPORT_TARGET_RELATIVE_SCALE),
                    TemplateCustomSemantics.draftInputForTargetType(
                            viewportInput, viewportTargetType,
                            TemplateConfigValue.VIEWPORT_TARGET_ABSOLUTE_DP),
                    fontScaleInput, fontApplyMode, selectedTypefaceId, fontHookDomainsRaw);
        }

        public Request(String templateId,
                String name,
                String viewportInput,
                String viewportTargetType,
                String viewportApplyMode,
                String viewportScaleInput,
                String viewportAbsoluteInput,
                String fontScaleInput,
                String fontApplyMode,
                String selectedTypefaceId,
                String fontHookDomainsRaw) {
            this.templateId = templateId != null ? templateId.trim() : null;
            this.name = name != null ? name.trim() : "";
            this.viewportInput = viewportInput != null ? viewportInput.trim() : "";
            this.viewportTargetType = TemplateConfigValue.normalizeViewportTargetType(viewportTargetType);
            this.viewportApplyMode = viewportApplyMode;
            this.viewportScaleInput = viewportScaleInput != null ? viewportScaleInput.trim() : "";
            this.viewportAbsoluteInput = viewportAbsoluteInput != null
                    ? viewportAbsoluteInput.trim()
                    : "";
            this.fontScaleInput = fontScaleInput != null ? fontScaleInput.trim() : "";
            this.fontApplyMode = fontApplyMode;
            this.selectedTypefaceId = selectedTypefaceId;
            this.fontHookDomainsRaw = fontHookDomainsRaw;
        }
    }

    public static final class Result {
        public final boolean success;
        public final int messageResId;
        public final String templateId;

        private Result(boolean success, int messageResId, String templateId) {
            this.success = success;
            this.messageResId = messageResId;
            this.templateId = templateId;
        }

        public static Result success(int messageResId, String templateId) {
            return new Result(true, messageResId, templateId);
        }

        public static Result failure(int messageResId, String templateId) {
            return new Result(false, messageResId, templateId);
        }
    }
}

File: app/src/main/java/com/dpis/module/templates/QuickTemplateSaveHandler.kt
package com.dpis.module.templates

import com.dpis.module.R
import java.util.LinkedHashSet

/** Validates and persists one quick-template editor submission. */
class QuickTemplateSaveHandler {
    fun save(store: QuickTemplateStore?, request: Request?): Result {
        if (store == null) return Result.failure(R.string.status_save_requires_init, null)
        if (request == null) return Result.failure(R.string.status_save_invalid, null)
        val name = request.name.trim()
        if (name.isEmpty()) return Result.failure(R.string.quick_template_name_required, request.templateId)
        if (store.hasDuplicateName(name, request.templateId)) {
            return Result.failure(R.string.quick_template_name_duplicate, request.templateId)
        }
        if (!TemplateCustomSemantics.isViewportInputValid(request.viewportInput, request.viewportTargetType) ||
            !TemplateCustomSemantics.isFontScaleInputValid(request.fontScaleInput)
        ) {
            return Result.failure(R.string.status_save_invalid, request.templateId)
        }

        val templateId = request.templateId?.takeUnless { it.isBlank() } ?: store.newTemplateId()
        val current = store.read(templateId)
        val selectedPackages = current?.selectedPackages ?: LinkedHashSet()
        val template = QuickTemplateStore.QuickTemplate(
            templateId,
            name,
            System.currentTimeMillis(),
            selectedPackages,
            TemplateCustomSemantics.fromEditorDraft(
                request.viewportInput,
                request.viewportTargetType,
                request.viewportApplyMode,
                request.viewportScaleInput,
                request.viewportAbsoluteInput,
                request.fontScaleInput,
                request.fontApplyMode,
                request.selectedTypefaceId,
                request.fontHookDomainsRaw
            )
        )
        return if (store.save(template)) {
            Result.success(R.string.quick_template_save_success, templateId)
        } else {
            Result.failure(R.string.quick_template_save_failed, templateId)
        }
    }

    class Request {
        @JvmField val templateId: String?
        @JvmField val name: String
        @JvmField val viewportInput: String
        @JvmField val viewportTargetType: String
        @JvmField val viewportApplyMode: String?
        @JvmField val viewportScaleInput: String
        @JvmField val viewportAbsoluteInput: String
        @JvmField val fontScaleInput: String
        @JvmField val fontApplyMode: String?
        @JvmField val selectedTypefaceId: String?
        @JvmField val fontHookDomainsRaw: String?

        constructor(
            templateId: String?,
            name: String?,
            viewportInput: String?,
            viewportTargetType: String?,
            viewportApplyMode: String?,
            viewportScaleInput: String?,
            viewportAbsoluteInput: String?,
            fontScaleInput: String?,
            fontApplyMode: String?,
            selectedTypefaceId: String?,
            fontHookDomainsRaw: String?
        ) {
            this.templateId = templateId?.trim()
            this.name = name?.trim().orEmpty()
            this.viewportInput = viewportInput?.trim().orEmpty()
            this.viewportTargetType = TemplateConfigValue.normalizeViewportTargetType(viewportTargetType)
            this.viewportApplyMode = viewportApplyMode
            this.viewportScaleInput = viewportScaleInput?.trim().orEmpty()
            this.viewportAbsoluteInput = viewportAbsoluteInput?.trim().orEmpty()
            this.fontScaleInput = fontScaleInput?.trim().orEmpty()
            this.fontApplyMode = fontApplyMode
            this.selectedTypefaceId = selectedTypefaceId
            this.fontHookDomainsRaw = fontHookDomainsRaw
        }

        constructor(
            templateId: String?,
            name: String?,
            viewportInput: String?,
            viewportTargetType: String?,
            viewportApplyMode: String?,
            fontScaleInput: String?,
            fontApplyMode: String?,
            selectedTypefaceId: String?,
            fontHookDomainsRaw: String?
        ) : this(
            templateId,
            name,
            viewportInput,
            viewportTargetType,
            viewportApplyMode,
            TemplateCustomSemantics.draftInputForTargetType(
                viewportInput, viewportTargetType, TemplateConfigValue.VIEWPORT_TARGET_RELATIVE_SCALE
            ),
            TemplateCustomSemantics.draftInputForTargetType(
                viewportInput, viewportTargetType, TemplateConfigValue.VIEWPORT_TARGET_ABSOLUTE_DP
            ),
            fontScaleInput,
            fontApplyMode,
            selectedTypefaceId,
            fontHookDomainsRaw
        )

    }

    class Result private constructor(
        @JvmField val success: Boolean,
        @JvmField val messageResId: Int,
        @JvmField val templateId: String?
    ) {
        companion object {
            @JvmStatic fun success(messageResId: Int, templateId: String?) = Result(true, messageResId, templateId)
            @JvmStatic fun failure(messageResId: Int, templateId: String?) = Result(false, messageResId, templateId)
        }
    }
}
